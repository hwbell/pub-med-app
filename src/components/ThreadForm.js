import React from 'react';
import '../App.scss';
import 'bootstrap/dist/css/bootstrap.css';

// components
import { Button, Modal, ModalHeader, ModalBody, ModalFooter, Input, Form } from 'reactstrap';

// ******************************************************************************
class ThreadForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      threadInfo: {},
      isNewThread: !this.props.thread
    }

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.renderInputs = this.renderInputs.bind(this);
  }

  handleChange(e) {

    let { name, value } = e.target;

    // use the target name as the key for the value in profileInfo
    let obj = this.state.threadInfo;
    obj[`${name}`] = value;

    this.setState({ threadInfo: obj }, () => {
      // console.log(this.state.user)
    })
  }

  handleSubmit() {

    this.props.handleSubmitThread(this.state.threadInfo);
    console.log(this.state.threadInfo)

    // wipe the state - we only need it before it is sent to the server
    this.setState({
      threadInfo: {}
    });

  }

  renderInputs() {
    // set the current value of the inputs from this.state.threadUpdates, if available.
    // this means we have changed something 
    let thread = this.props.thread;
    let threadUpdates = this.state.threadInfo;

    let threadProps = ['name', 'article', 'paragraph'];
    
    let placeholders = {
      name: 'Give your post a clear, searchable name!',
      article: 'Include the PMID or PMCID of related articles.',
      paragraph: 'Give a short description of the topic of your thread.',
    }

    let threadInfo = {};

    threadProps.forEach((prop) => {

      if (threadUpdates[prop]) {
        threadInfo[prop] = threadUpdates[prop];
      } else if (thread && thread[prop]) {
        threadInfo[prop] = thread[prop];
      } else {
        threadInfo[prop] = ''
      }

    })

    return Object.keys(threadInfo).map((name, i) => {
      console.log(placeholders[name])

      console.log(i, name)

      let title = name[0].toUpperCase() + name.slice(1);
      if (title === 'Name' || title === 'Paragraph') {
        title += ` *`
      }

      let type = 'text';
      let style = styles.input;
      if (name === 'paragraph' || name === 'comment') {
        type = 'textarea';
        style = styles.inputLarge;
      }

      return (
        <div key={i}>
          <p style={{ color: 'black' }}>{title}</p>
          <Input style={style} type={type} name={name}
            maxLength={500}
            value={threadInfo[name] || ''}
            placeholder={placeholders[name]}
            onChange={(e) => this.handleChange(e)}
          />
        </div>
      )

    })
  }

  render() {

    return (
      <Modal centered={true}
        isOpen={this.props.isVisible}
        toggle={this.props.toggle}>

        <ModalHeader>Add some info to your profile</ModalHeader>
        <ModalBody>

          <Form style={styles.form} onSubmit={this.handleSubmit}>

            {this.renderInputs()}

          </Form>

        </ModalBody>
        <ModalFooter>
          <Button color="primary" size="sm" onClick={this.handleSubmit}>save</Button>
          <Button color="secondary" size="sm" onClick={this.props.toggle}>cancel</Button>
        </ModalFooter>

      </Modal>
    );
  }
}

const styles = {
  input: {
    marginBottom: '20px',
    textAlign: 'left'
  },
  inputLarge: {
    marginBottom: '20px',
    height: '140px',
    textAlign: 'left',
    paddingTop: 0
  }
}

export default ThreadForm;
