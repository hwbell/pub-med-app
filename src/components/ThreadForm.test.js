import React from 'react';
import ThreadForm from './ThreadForm';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

// stub for toggling
const toggleStub = jest.fn();
const handleSubmitThreadStub = jest.fn();
const handleChangeStub = jest.fn();

const thread = {
  name: 'Sourcing of PMID: 013091283',
  article: '013091283',
  user: 'The Creator',
  paragraph: 'I found the source material for the basis of this study to be lacking. Does anyone else agree?',
}
const someProps = {
  thread,
  isVisible: false,
  toggle: toggleStub,
  handleSubmitThread: handleSubmitThreadStub,
  handleChange: handleChangeStub
}

describe('ThreadForm', () => {

  let wrapper, instance;

  beforeEach(() => {
    wrapper = shallow(<ThreadForm {...someProps} />)
    instance = wrapper.instance();
  })

  // tests
  it('renders without crashing', async () => {
    wrapper.update();
  });

  it('renders correctly', async () => {
    const tree = await renderer
      .create(<ThreadForm />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('contains the correct elements', () => {

    ['Modal', 'ModalBody', 'ModalFooter', 'Form',{ color: "primary" }, { color: "primary" }].forEach((selector) => {
      expect(wrapper.find(selector).length).toEqual(1);
    });

    // one input for each property of the thread
    expect(wrapper.find('Input').length).toBe(3);

  });

  it('should have the correct placeholder values', () => {
    expect(wrapper.find('Input').at(0).props().placeholder).toBe("Give your post a clear, searchable name!");
    expect(wrapper.find('Input').at(1).props().placeholder).toBe("Include the PMID or PMCID of related articles.")
    expect(wrapper.find('Input').at(2).props().placeholder).toBe("Give a short description of the topic of your thread.")

  })

  it('should use values from this.props.user if there is not threadInfo in state', () => {
    expect(wrapper.find('Input').at(0).props().value).toBe(thread.name);
    expect(wrapper.find('Input').at(1).props().value).toBe(thread.article)
    expect(wrapper.find('Input').at(2).props().value).toBe(thread.paragraph)
  })

  it('should have an empty threadInfo in state to start', () => { 
    expect(wrapper.state().threadInfo).toMatchObject({})
  })

  it('should save threadInfo in state when user types in info', () => { 
    let e = {
      target: {
        name: 'about',
        paragraph: 'This article seemed unreliable, but upon futher reading ... '
      }
    }
    wrapper.instance().handleChange(e)
    wrapper.update();

    expect(wrapper.state().threadInfo).toMatchObject({
      about: e.paragraph
    })
  })

  it('should fire the provided handleSumbitThread() function when OK is clicked', () => {

    wrapper.find({color: 'primary'}).simulate('click');
    wrapper.update();

    expect(handleSubmitThreadStub.mock.calls.length).toBe(1)
    expect(handleSubmitThreadStub).toBeCalledWith(wrapper.state().threadInfo)
  })

  it('should fire the provided toggle when the cancel button is clicked', () => {

    wrapper.find({ color: "secondary" }).simulate('click');
    expect(toggleStub.mock.calls.length).toBe(1);

  })

})