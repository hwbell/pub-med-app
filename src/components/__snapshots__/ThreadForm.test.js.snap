// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreadForm renders correctly 1`] = `null`;
